src/components/molecules/ProfilePictureInput/ProfilePictureInput.tsx

```diff
@@ -42,6 +44,16 @@ const ProfilePictureInput: React.FunctionComponent<PropsType> = ({
   const [error, setError] = useState("");
   const firebase = useFirebase();
   const uploadRef = useRef<HTMLInputElement>(null);
+  const queryParams = useQuery();
+  const venueId = queryParams.get("venueId");
```

Comment:
Instead of reading from URL params this deep in the code, how much effort would be required to pass the `venueId` as a prop?

Does this actually need to be the `venueId` here? It would make more sense to use the `sovereignVenueId` I would think? As that correlates closer to the concepts in the new world architecture.

Reply:
> Does this actually need to be the venueId here? It would make more sense to use the sovereignVenueId I would think? As that correlates closer to the concepts in the new world architecture.

I'm not 100% sure on this one. `sovereignVenueId` is probably a better way, because the avatars can be set only for the parent venue and all of it's children will get the avatars from there.

Reply:
Cool, let's go with that then. Which means we can use `useSovereignVenueId` (which seemingly currently just reads from `useVenueId` anyway.. _sigh_ :p)

Reply:
`sovereignVenueId` is now fetched and used in https://github.com/sparkletown/sparkle/pull/1409/commits/6cdbdb2d81f61c67a0665f91bf358d09fc89bfbe however, the defualt avatars are visible for a split second between the 2 async requests. Any idea on how to fix this? Maybe use just one state for loading and change it manually on `.finally()`

Reply:
Looks like you're using `fetchSovereignVenueId` (the api layer) in that commit, whereas I said to use `useSovereignVenueId` (the hook that wraps up this functionality and is what we should be using from the component layer) from `src/hooks/useSovereignVenueId.ts`

It stores state in redux, so you may never even need to wait on it, but on the off chance that you do, it returns `isSovereignVenueIdLoading`

That should be combined into a single loading state for both (which it seems you're already basically doing with `isLoading`)

The `avatarImages` `useMemo` should probably have a guard that just returns an empty array while `isLoading` is true, since there's no point to processing/rendering it if we're still loading.

Then in the JSX since it's such a small amount of code now, it probably makes sense to use a ternary.

```jsx
{isLoading ? <div>Loading...</div> : avatarImages}
```

---

> however, the defualt avatars are visible for a split second between the 2 async requests

Is this because `useAsync` doesn't actually 'fire' until the first load/display? If so, we could potentially look at it's return value (assuming it is `undefined` or similar before the hook has run?), and include that in the `isLoading` variable?

Something like:

```typescript
const isLoading = (isLoadingSovereignVenueId || isLoadingCustomAvatars) && customAvatars !== undefined;
```

Reply:
> Looks like you're using fetchSovereignVenueId (the api layer) in that commit, whereas I said to use useSovereignVenueId (the hook that wraps up this functionality and is what we should be using from the component layer) from src/hooks/useSovereignVenueId.ts

`useSovereignVenueId` is only working with params, where in my case the url has search query. It returns `undefined` and that's why I had to use `fetchSovereignVenueId`

Reply:
I fixed the loading issue in: https://github.com/sparkletown/sparkle/pull/1409/commits/d9d670ae3e2af69dca89fb490b8b02209ae9baa4

Reply:
> `useSovereignVenueId` is only working with params, where in my case the url has search query. It returns `undefined` and that's why I had to use `fetchSovereignVenueId`

Makes sense. That is changed in https://github.com/sparkletown/sparkle/pull/1368 (but not merged yet)

- https://github.com/sparkletown/sparkle/pull/1368/files?file-filters%5B%5D=.ts&hide-deleted-files=true#diff-972c4ca9a9c40bfadf97da41744c77a76d472da0c96d80ed230fd2b216f8057e

Reply:
> I fixed the loading issue in: [d9d670a](https://github.com/sparkletown/sparkle/commit/d9d670ae3e2af69dca89fb490b8b02209ae9baa4)

Sweet, if it works, go with that. But also probably add an `@debt` comment saying to use `useSovereignVenueId` once it's refactored to accept `venueId` as a param.

Reply:
@0xdevalias I will add the @debt in a moment. Meanwhile, can you please review this one final time because Sofi needs this in like an hour.

Reply:
`@debt` has been added in: https://github.com/sparkletown/sparkle/pull/1409/commits/82c66b6da26694ba0cf53c43496e1d22742033b9

I had to commit some additional changes, `venueId` is now passed as a prop because this component is used in multiple places and `useVenueId()` and `useQuery()` are both not suitable. Sometimes the url has search query, sometimes it's param. https://github.com/sparkletown/sparkle/pull/1409/commits/843431ffa592ffb19a91978deb4fec4425982c17

Reply:
Makes sense.

Yeah, I can review it in a sec.